nextmv/cloud/package.py

```diff
@@ -0,0 +1,370 @@
+
+    if verbose:
+        log(result.stdout)
+
+
+def _run_pre_push_command(
+    app_dir: str,
+    pre_push_command: Optional[str] = None,
+    verbose: bool = False,
+) -> None:
+    """Run the pre-push command specified in the manifest."""
+
+    if pre_push_command is None or pre_push_command == "":
+        return
+
+    elements = ["bash", "-c", pre_push_command]
+    if platform.system() == "Windows":
+        elements = ["cmd", "/c", pre_push_command]
+
+    command_str = " ".join(elements)
+    log(f'🔨 Running pre-push command: "{command_str}"')
+    try:
+        result = subprocess.run(
+            elements,
+            env=os.environ,
+            check=True,
+            text=True,
+            capture_output=True,
+            cwd=app_dir,
+        )
+
+    except subprocess.CalledProcessError as e:
+        raise Exception(f"error running pre-push command: {e.stderr}") from e
+
+    if verbose:
+        log(result.stdout)
+
+
+def __find_files(
+    app_dir: str,
+    filters: List[str],
+) -> Tuple[List[str], List[str], List[Dict[str, str]]]:
+    """Find all files matching the given filters in the given directory."""
+
+    found = []
+    missing = []
+
+    # Temporarily switch to the directory to make the globbing work
+    cwd = os.getcwd()
+    try:
+        os.chdir(app_dir)
+    except OSError as e:
+        raise Exception(f"error changing to file root directory: {e}") from e
+
+    for filter in filters:
+        # We support "some/path/" ending with a "/". We consider it equivalent
+        # to "some/path/*".
+        pattern = filter
+        if filter.endswith("/"):
+            pattern = filter + "*"
+        # If the pattern starts with a '!': negate the pattern
+        negated = False
+        if pattern.startswith("!"):
+            pattern = pattern[1:]
+            negated = True
+        matches = glob.glob(pattern)
+        if not matches and not negated:
+            missing.append(filter)
+        else:
+            if negated:
+                found = [f for f in found if f not in matches]
+            else:
+                found.extend(matches)
+
+    # Switch back to the original directory
+    os.chdir(cwd)
+
+    files = []
+    for file in found:
+        files.append(
+            {
+                "interior_path": file,
+                "absolute_path": os.path.join(app_dir, file),
+            }
+        )
+
+    return found, missing, files
+
+
+def __confirm_mandatory_files(manifest: Manifest, present_files: List[str]) -> None:
+    """Confirm that all mandatory files are present in the given list of files."""
+
+    mandatory_files = _MANDATORY_FILES_PER_TYPE[manifest.type]
+    found_files = {os.path.normpath(file): True for file in present_files}
+    missing_files = [file for file in mandatory_files if file not in found_files]
+
+    if missing_files:
+        raise Exception(f"missing mandatory files: {', '.join(missing_files)}")
+
+
+def __install_dependencies(
```

Comment:
I would highly recommend evaluating the possibility to invoke `pip` directly, as we are already _within_ Python here. Something like this should do:

```python
import pip


def install(package):
    pip.main(["install", package])


# Example
if __name__ == "__main__":
    install("requests")
```

However, as I haven't confirmed that this works in all situations, I won't block. It's just that relying on a correct `PATH` setup, will fail us at some point (imho).

Reply:
(it may also simplify the code)

Reply:
Then again, this may not be safe way to invoke pip. :thinking:

Reply:
From the [official docs](https://pip.pypa.io/en/stable/user_guide/#using-pip-from-your-program):
> While it is implemented in Python, and so is available from your Python code via import pip, you must not use pip’s internal APIs in this way

So let’s leave it the way it is 🙏🏻 .

Reply:
Yeah, saw that then as well. A little unfortunate as we then need to rely on local setup being proper, but fair enough. :relaxed: